feat(assure_command_shell): assure proper installation of a shell

FGrose opened the pull request:
Introduce a new meta module, 99~sh, that will check that a command shell has been specified after module checks have been performed and before executable dependencies are checked.  This avoids implicit installations of a command shell that skip the dracut module checks that the shell might need.

This PR now includes 5 commits to implement the feature:
1. feat(module-postprocess): provide a postprocessing loop for modules
2. feat(99~sh): assure explicit installation of a command shell 
3. feat(99~no&#126;sh): allow a build without any command shell
4. fix(00bash): fail if Bash below version 4.4 is linked to /bin/sh 
5. refactor(99squash): gather code into the squash module

2 (indirectly) and 3, 4, & 5 above use the postprocessing feature directly to achieve the intention.


Completes fix to #1530 augmenting #1534.

An alternative command shell, for example, _zsh_ can be installed by providing an appropriate `module-setup.sh` at `../modules.d/00zsh/` and running dracut with `ln -sf zsh /bin/sh`.

### Checklist
- [✔] I have tested it locally
-  [✔] I have reviewed and updated documentation in dracut.modules.7.asc & HACKING.md

FGrose commented:
@LaszloGombos  > busybox is also supported as a shell
Updated to include busybox in supported list.

LaszloGombos reviewed (commented):
Overall direction is great, but needs some cleanup

FGrose commented:
@LaszloGombos re: `/bin/sh` not found
See PR #2394.

Henrik66 commented:
@FGrose Do you plan on working on this PR more based on feedback from @LaszloGombos ? I am asking because I am also available try to help with this PR. Thank you @FGrose .

FGrose commented:
> @FGrose Do you plan on working on this PR more based on feedback from @LaszloGombos ? I am asking because I am also available try to help with this PR. Thank you @FGrose .

@Henrik66 - Yes, I have a refactoring that I am testing now.

FGrose committed:
feat(module-postprocess): provide a postprocessing loop for modules

Provide code for postprocessing the build after all modules have been
loaded by enrolling those modules with a postprocess() function in
their module_setup.sh into the variable $mods_to_postprocess. Update
the documentation in dracut.modules.7.asc and HACKING.md files.

Also, include a missing dinfo message before including modules.

This feature is demonstrated in three upcoming commits:
  feat(99~no~sh): allow a build without any command shell
  fix(00bash): fail if Bash below version 4.4 is linked to /bin/sh
  refactor(99squash): gather code into the squash module

FGrose committed:
feat(99~sh): assure explicit installation of a command shell

Introduce a new meta module '99~sh' and assure that its depends()
function runs at the end of the 'for_each_module_dir()' loop
by using LC_COLLATE=C in that function so that the '99~sh' meta
module sorts after other normally named modules.

Background:
Implicit installation of command shells may occur during the ***
Resolving executable dependencies *** phase of dracut.sh with the
default flag $DRACUT_RESOLVE_LAZY=1.  Any module script file with
a shebang interpreter directive will trigger the installation of
that command interpreter, typically /bin/sh, unless that
interpreter has already been installed.  Such implicit
installations bypass any conditions encoded in their dracut
module dependency checks and should be precluded.

If the dracut -m, --modules option is used (outside of -m 'auto'),
no module-setup.sh level code is invoked during the
'for_each_module_dir check_module' loop for any modules other than
those specified in the -m option arguments. For this reason, in
this case invoke module_depends ~sh.

To allow an initramfs to be built without any command shell, as
currently allowed (such as when the -m option is used or when
incrementally building the initramfs with --rebuild), a new meta
module '99~no~sh' will be introduced in a following commit.

FGrose committed:
feat(99~no~sh): allow a build without any command shell

Install a null link to /bin/sh in order to satisfy the executable
dependency checks and remove it in a postprocess() function.

A no-command-shell-build may be desired when the -m, --modules
option is used or when incrementally building an initramfs
with --rebuild.

This module provides an opt-out to the automatic specification
of a command shell introduced with 99~sh.

FGrose committed:
refactor(99squash): gather code into the squash module

Move squash-module-specific code from dracut.sh to
/99squash/module-setup.sh.
Use dracut's new module_postprocess() feature.

Make shell installation explicit, preferring busybox, if available.

Clarify a comment about the dracut directory files.

Also, make new shellcheck adjustments.

FGrose committed:
fix(00bash): fail if Bash below version 4.4 is linked to /bin/sh

At least bash-4.4 (from 2016-09-15) is required for proper xtrace
logging when Bash is the initramfs command shell. This requirement
was introduced in commit 10cf8e4 on 2022-12-30 with the feature,
local -, which automates the restoration of local xtrace & other
set options.

FGrose commented on modules.d/00_sh/module-setup.sh:
```diff
@@ -0,0 +1,33 @@
+    # Assure a command shell.
+    [[ " $add_dracutmodules $force_add_dracutmodules " == @(*\ dash\ *|*\ bash\ *|*\ mksh\ *|*\ busybox\ *) ]] || {
+        shells='dash bash mksh busybox'
+        _shell=$(realpath -e /bin/sh)
```
The refactored PR meets the above intentions with requiring the user to request the meta module. It operates automatically.

FGrose commented on modules.d/00_sh/module-setup.sh:
```diff
@@ -0,0 +1,33 @@
+
+# Prerequisite check(s) for module.
+check() {
+
```
The refactored PR respects the shell requests of other modules.

FGrose commented on modules.d/00_sh/module-setup.sh:
```diff
@@ -0,0 +1,33 @@
+}
+
+# Module dependency requirements.
+depends() {
```
'depends() function should return the shell selected.' This is achieved if needed.
The 99base module only expects that some shell be installed and linked to `/bin/sh`.  This is achieved.

FGrose commented on modules.d/00bash/module-setup.sh:
```diff
@@ -24,8 +24,21 @@ depends() {
+
     inst /bin/bash
 
+    # local - (available since bash-4.4 2016-09-15) automates the
```
00bash and 99squash module changes are included in separate commits.
They are needed to accomplish the intention of installing the command shell only after all installation conditions have been met.

FGrose commented on modules.d/00_sh/module-setup.sh:
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
```
The module name '99~sh' is required for proper ordering of the code.
It is a meta module enforcing some&#126;shell—not to be confused with any current sh.

FGrose commented on modules.d/99base/module-setup.sh:
```diff
@@ -24,11 +24,6 @@ install() {
-        (ln -s bash "${initdir}/bin/sh" || :)
-    fi
-
     # add common users in /etc/passwd, it will be used by nfs/ssh currently
```
This is not needed with the refactored approach.

FGrose commented on modules.d/80test-root/module-setup.sh:
```diff
@@ -22,6 +22,7 @@ install() {
     inst_binary "${dracutbasedir}/dracut-util" "/usr/bin/dracut-util"
     ln -s dracut-util "${initdir}/usr/bin/dracut-getarg"
     ln -s dracut-util "${initdir}/usr/bin/dracut-getargs"
+    [[ -L /bin/sh ]] || ln -sf dash /bin/sh
```
This seems to have been resolved.
